lib/auth/register.go

```diff
@@ -109,37 +109,51 @@ func Register(params RegisterParams) (*Identity, error) {
 		return nil, trace.Wrap(err)
 	}
 
-	// Attempt to register through the auth server, if it fails, try and
-	// register through the proxy server.
-	ident, err := registerThroughAuth(token, params)
-	if err != nil {
-		// If no params client was set this is a proxy and fail right away.
-		if params.GetHostCredentials == nil {
-			log.Debugf("Missing client, failing with error from Auth Server: %v.", err)
-			return nil, trace.Wrap(err)
-		}
-
-		log.Errorf("Failed to register through auth server: %v; falling back to trying the proxy server", err)
+	type registerMethod struct {
+		call func(token string, params RegisterParams) (*Identity, error)
+		desc string
+	}
+	registerThroughAuth := registerMethod{registerThroughAuth, "with auth server"}
+	registerThroughProxy := registerMethod{registerThroughProxy, "via proxy server"}
+
+	var registerMethods []registerMethod
+	if params.GetHostCredentials == nil {
+		log.Debugf("Missing client, it is not possible to register through proxy.")
+		registerMethods = []registerMethod{registerThroughAuth}
+	} else if authServerIsProxy(params) {
+		log.Debugf("The first specified auth server appears to be a proxy.")
+		registerMethods = []registerMethod{registerThroughProxy, registerThroughAuth}
+	} else {
+		registerMethods = []registerMethod{registerThroughAuth, registerThroughProxy}
+	}
 
-		// params.AuthServers could contain a proxy address, to deal with nodes
-		// behind NAT. Try registering using the proxy API.
-		ident, err = registerThroughProxy(token, params)
+	var collectedErrs []error
+	for i, method := range registerMethods {
+		log.Debugf("Attempting registration %s.", method.desc)
+		ident, err := method.call(token, params)
 		if err != nil {
-			return nil, trace.Wrap(err, "failed to register through proxy server: %v", err)
+			collectedErrs = append(collectedErrs, err)
+			log.WithError(err).Debugf("Registration %s failed.", method.desc)
+			if i+1 < len(registerMethods) {
+				log.Infof("Unable to register %s, falling back to registration %s.",
+					method.desc, registerMethods[i+1].desc)
+			}
+		} else {
+			log.Infof("Successfully registered %s.", method.desc)
+			return ident, nil
 		}
-
-		log.Debugf("Successfully registered through proxy server.")
-		return ident, nil
 	}
+	return nil, trace.NewAggregate(collectedErrs...)
+}
 
-	log.Debugf("Successfully registered through auth server.")
-	return ident, nil
+// authServerIsProxy returns true if the first specified auth server
+// to register with appears to be a proxy.
+func authServerIsProxy(params RegisterParams) bool {
```

Comment:
Provide as little context as necessary for this to work - e.g. instead of the whole params value, pass just the servers slice.